File: UTFPR/Sistemas Operacionais/CCH/Filosofos.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <unistd.h> // para a função sleep

#define NUM_FILOSOFOS 5

pthread_mutex_t garfo[NUM_FILOSOFOS]; // Mutexes para os garfos

// Função que simula o filósofo comendo
void* comer(void* id) {
    int filosofo_id = *((int*)id);

    while (1) {
        // Tenta pegar os garfos (mutexes)
        int garfoEsquerdo = filosofo_id;
        int garfoDireito = (filosofo_id + 1) % NUM_FILOSOFOS;

        printf("Filósofo %d está tentando comer...\n", filosofo_id + 1);

        pthread_mutex_lock(&garfo[garfoEsquerdo]);  // Pega o garfo da esquerda
        pthread_mutex_lock(&garfo[garfoDireito]);   // Pega o garfo da direita

        // Comer
        printf("Filósofo %d está comendo.\n", filosofo_id + 1);
        sleep(1);  // Simula o tempo de comer

        // Libera os garfos (mutexes)
        pthread_mutex_unlock(&garfo[garfoEsquerdo]);
        pthread_mutex_unlock(&garfo[garfoDireito]);

        printf("Filósofo %d terminou de comer.\n", filosofo_id + 1);

        // Pensar
        printf("Filósofo %d está pensando...\n", filosofo_id + 1);
        sleep(2);  // Simula o tempo de pensar
    }

    return NULL;
}

int main() {
    pthread_t filosofos[NUM_FILOSOFOS];
    int ids[NUM_FILOSOFOS];

    // Inicializa os mutexes para os garfos
    for (int i = 0; i < NUM_FILOSOFOS; i++) {
        pthread_mutex_init(&garfo[i], NULL);
    }

    // Cria os threads para os filósofos
    for (int i = 0; i < NUM_FILOSOFOS; i++) {
        ids[i] = i;
        pthread_create(&filosofos[i], NULL, comer, &ids[i]);
    }

    // Espera todos os filósofos terminarem (teoricamente, nunca vai acontecer)
    for (int i = 0; i < NUM_FILOSOFOS; i++) {
        pthread_join(filosofos[i], NULL);
    }

    // Destrói os mutexes
    for (int i = 0; i < NUM_FILOSOFOS; i++) {
        pthread_mutex_destroy(&garfo[i]);
    }

    return 0;
}

File: UTFPR/Sistemas Operacionais/CCH/Filosofos.js
class Garfo {
    constructor(id) {
        this.id = id;
        this.disponivel = true;
    }

    async pegar() {
        while (!this.disponivel) {
            await new Promise(resolve => setTimeout(resolve, 100));
        }
        this.disponivel = false;
    }

    async largar() {
        this.disponivel = true;
    }
}

class Filosofo {
    constructor(id, garfoEsquerda, garfoDireita) {
        this.id = id;
        this.garfoEsquerda = garfoEsquerda;
        this.garfoDireita = garfoDireita;
    }

    async comer() {
        while (true) {
            const tempoPensando = Math.random() * 1000;
            console.log(`[${new Date().toISOString()}] Filósofo ${this.id} está pensando...`);
            await new Promise(resolve => setTimeout(resolve, tempoPensando));

            console.log(`[${new Date().toISOString()}] Filósofo ${this.id} quer comer...`);
            await this.garfoEsquerda.pegar();
            console.log(`[${new Date().toISOString()}] Filósofo ${this.id} pegou o garfo da esquerda (${this.garfoEsquerda.id})`);

            await this.garfoDireita.pegar();
            console.log(`[${new Date().toISOString()}] Filósofo ${this.id} pegou o garfo da direita (${this.garfoDireita.id})`);

            console.log(`[${new Date().toISOString()}] Filósofo ${this.id} está comendo com os garfos (${this.garfoDireita.id}) e (${this.garfoEsquerda.id})...`);
            const tempoComendo = Math.random() * 1000;
            await new Promise(resolve => setTimeout(resolve, tempoComendo));

            await this.garfoDireita.largar();
            console.log(`[${new Date().toISOString()}] Filósofo ${this.id} devolveu o garfo da direita (${this.garfoDireita.id})`);

            await this.garfoEsquerda.largar();
            console.log(`[${new Date().toISOString()}] Filósofo ${this.id} devolveu o garfo da esquerda (${this.garfoEsquerda.id})`);
        }
    }
}

const numFilosofos = 5;
const garfos = Array.from({ length: numFilosofos }, (_, i) => new Garfo(i));
const filosofos = garfos.map((garfo, i) => new Filosofo(i, garfo, garfos[(i + 1) % numFilosofos]));

async function iniciarJantar() {
    const promessasJantar = filosofos.map(filosofo => filosofo.comer());
    await Promise.all(promessasJantar);
}

iniciarJantar();